test/Microsoft.AspNetCore.Routing.Tests/RouteCollectionTest.cs

```diff
@@ -363,6 +363,79 @@ public static IEnumerable<object[]> IntegrationTestData
             Assert.Empty(pathData.DataTokens);
         }
 
+        public static TheoryData<RestoreRouteDataVariation> RestoresRouteDataForEachRouterData
+        {
+            get
+            {
+                var data = new TheoryData<RestoreRouteDataVariation>();
+
+                // Here 'area' segment doesn't have a value but the later segments have values. This is an invalid
+                // route match and the url generation should look into the next available route in the collection.
+                var variation = new RestoreRouteDataVariation();
+                variation.Routes.Add(new RouteInfo("1", "{area?}/{controller=Home}/{action=Index}/{id?}"));
+                variation.Routes.Add(new RouteInfo("2", "{controller=Home}/{action=Index}/{id?}"));
+                variation.Values = new RouteValueDictionary(new { controller = "Test", action = "Index" });
+                variation.ExpectedUrl = "/Test";
+                variation.ExpectedRouteToMatch = "2";
+                data.Add(variation);
+
+                variation = new RestoreRouteDataVariation();
+                variation.Routes.Add(new RouteInfo("1", "{category:int}/{controller=Home}/{action=Index}/{id?}"));
+                variation.Routes.Add(new RouteInfo("2", "{controller=Home}/{action=Index}/{id?}"));
+                variation.Values = new RouteValueDictionary(new { category = "nan", controller = "Test", action = "Index" });
+                variation.ExpectedUrl = "/Test?category=nan";
+                variation.ExpectedRouteToMatch = "2";
+                data.Add(variation);
+
+                // Empty segment is an invalid segment and so route "1" cannot be used for generation
+                variation = new RestoreRouteDataVariation();
+                variation.Routes.Add(new RouteInfo("1", "{category}/{controller=Home}/{action=Index}/{id?}"));
+                variation.Routes.Add(new RouteInfo("2", "{controller=Home}/{action=Index}/{id?}"));
+                variation.Values = new RouteValueDictionary(new { category = "", controller = "Test", action = "Index" });
+                variation.ExpectedUrl = "/Test";
+                variation.ExpectedRouteToMatch = "2";
+                data.Add(variation);
+
+                // Here the segment 'a' is valid but 'b' is not as it would be empty. This would be an invalid route match, but
+                // the route value of 'a' should still be present to be evaluated for the next available route.
+                variation = new RestoreRouteDataVariation();
+                variation.Routes.Add(new RouteInfo("1", "{a}/{b?}/{c}"));
+                variation.Routes.Add(new RouteInfo("2", "{a=Home}/{b=Index}"));
+                variation.Values = new RouteValueDictionary(new { a = "Test", c = "Foo" });
+                variation.ExpectedUrl = "/Test?c=Foo";
+                variation.ExpectedRouteToMatch = "2";
+                data.Add(variation);
+
+                return data;
+            }
```

Comment:
I added a new TemplateBinder test and simplified the existing tests. I didn't remove the existing tests because they are integration tests and more easy to follow through as a scenario than the targeted template binder test.